pkg/plugins/steady/steady.go

```diff
@@ -0,0 +1,74 @@
+package steady
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	clusterapiv1 "open-cluster-management.io/api/cluster/v1"
+	clusterapiv1alpha1 "open-cluster-management.io/api/cluster/v1alpha1"
+	"open-cluster-management.io/placement/pkg/plugins"
+)
+
+const (
+	placementLabel = "cluster.open-cluster-management.io/placement"
+	description    = `
+	Steady prioritizer ensure the existing decision is stabilized. The clusters that existing decisions
+	choose are given the highest score while the clusters with no existing decisions are given the lowest
+	score.
+	`
+)
+
+var _ plugins.Prioritizer = &Steady{}
+
+type Steady struct {
+	handle plugins.Handle
+}
+
+func New(handle plugins.Handle) *Steady {
+	return &Steady{handle: handle}
+}
+
+func (s *Steady) Name() string {
+	return "steady"
+}
+
+func (s *Steady) Description() string {
+	return description
+}
+
+func (s *Steady) Score(
+	ctx context.Context, placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) (map[string]int64, error) {
+	// query placementdecisions with label selector
+	scores := map[string]int64{}
+	requirement, err := labels.NewRequirement(placementLabel, selection.Equals, []string{placement.Name})
+
+	if err != nil {
+		return nil, err
+	}
+
+	labelSelector := labels.NewSelector().Add(*requirement)
+	decisions, err := s.handle.DecisionLister().PlacementDecisions(placement.Namespace).List(labelSelector)
+
+	if err != nil {
+		return nil, err
+	}
+
+	existingDecisions := sets.String{}
+	for _, decision := range decisions {
+		for _, d := range decision.Status.Decisions {
+			existingDecisions.Insert(d.ClusterName)
+		}
+	}
+
+	for _, cluster := range clusters {
+		if existingDecisions.Has(cluster.Name) {
+			scores[cluster.Name] = plugins.MaxClusterScore
+		} else {
+			scores[cluster.Name] = 0
```

Comment:
Why the clustes not included in existing decisions got less score?

In my opinion, the clusters not included in existing decisions are new ones from new cluster set. These clusters should have more priorities for use. Please correct me if I am wrong.

Reply:
this is to ensure that the decision that has been made will not be churned. For instance, if cluster1 is already in the decision list, we will still prefer cluster1 in this prioritizer.

Reply:
OK. This will keep scheduled cluster not removed from existing decison in case of low score when to update the decision.
I am not sure to what extent this will keep the scheduled cluster safe. Because this score may fluctuated time to time.  At this point, some ideas come into my mind. When to update existing decisions, there should be some rules, an important one may be liveness of the cluster. So regarding the updating of the existing decisions, we may not depend on the score but some rules. The score should be only used when to schedule/create decisions.
Once decision done, unless some sever issue, we may not want to update it.  Just some my thoughts.

Reply:
I think that falls into filter parts. The steady only affect after filter. But if a cluster is unhealthy or unreachable, the filters should exclude it beforehand.